### PR TITLE
sockets

### PR DESCRIPTION
**Pull request by tanner0101**
- add `SecurityLayer` here
- add new `EngineClient` and `EngineServer`

Waiting on https://github.com/vapor/engine/pull/115

### CODE REVIEW

File: Sources/Vapor/Droplet/Droplet+TLS.swift [hunk 1]
```diff
@@ -5,9 +5,9 @@ extension Droplet {
     internal func makeServerConfig() throws -> ServerConfig {
         let serverConfig = config["server"]
         let port = serverConfig?["port"]?.int ?? cliPort(arguments: arguments) ?? 8080
-        let host = serverConfig?["host"]?.string ?? "0.0.0.0"
+        let hostname = serverConfig?["host"]?.string ?? "0.0.0.0"
         let securityLayer = try makeSecurityLayer(serverConfig: serverConfig)
-        return ServerConfig(host: host, port: port, securityLayer: securityLayer)
+        return ServerConfig(hostname: hostname, port: port.port, securityLayer)
```
Comment (LoganWright) [inline Sources/Vapor/Droplet/Droplet+TLS.swift]: what is `port.port`? things with property with same name that is different thing are always confusing to me

File: Sources/Vapor/Engine/ClientProtocol.swift [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+extension ClientProtocol {
+    /// Creates a new client from the information in the 
+    /// Request URI and uses it to respond to the request.
+    public static func respond(
+        to req: Request,
+        _ s: SecurityLayer? = nil
+    ) throws -> Response {
+        // use security layer from input or
+        // determine based on req uri scheme
+        let securityLayer = try s ??
```
Comment (LoganWright) [inline Sources/Vapor/Engine/ClientProtocol.swift]: this line is a little long w/ `??` and ternary nested, also needs to account for `wss`

let securityLayer: SecurityLayer
if let s = s {
    securityLayer = s
} else if request.uri.scheme.isSecure {
    securityLayer = default
} else {
    securityLayer = .none
}
```

File: Sources/Vapor/Engine/ClientProtocol.swift [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            ((req.uri.scheme ==  "https")
+                ? .tls(try EngineClient.defaultTLSContext())
+                : .none)
+
+        let client = try Self.init(
+            hostname: req.uri.hostname,
+            port: req.uri.port ?? 80,
```
Comment (LoganWright) [inline Sources/Vapor/Engine/ClientProtocol.swift]: if secure, default to 443

File: Sources/Vapor/Engine/ClientProtocol.swift [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        if let body = body {
+            req.body = body.makeBody()
+        }
+        return try respond(to: req)
+    }
+}
+
+// MARK: TLS
+
+private var _defaultTLSClientContext: () throws -> (TLS.Context) = {
```
Comment (LoganWright) [inline Sources/Vapor/Engine/ClientProtocol.swift]: I don't get it? Is this some sort of generic workaround for why we can't use getter setter below? If so, quick comment

File: Sources/Vapor/Engine/ServerProtocol.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    /// to respond to accepted requests
+    func start(
+        _ responder: Responder,
+        errors: @escaping ServerErrorHandler
+    ) throws
+}
+
+// MARK: TLS
+
+private var _defaultTLSServerContext: () throws -> (TLS.Context) = {
```
Comment (LoganWright) [inline Sources/Vapor/Engine/ServerProtocol.swift]: same questino here from above

File: Tests/VaporTests/DropletTests.swift [hunk 6]
```diff
@@ -86,7 +86,7 @@ class DropletTests: XCTestCase {
 
         drop.console.wait(seconds: 1)
 
-        let res = try drop.client.get("http://0.0.0.0:8523/foo")
+        let res = try drop.client.request(.get, "http://0.0.0.0:8523/foo")
```
Comment (LoganWright) [inline Tests/VaporTests/DropletTests.swift]: has the old syntax been removed, I feel very strongly about maintaining the old syntax support. Even if it's in addition to this.

File: Tests/VaporTests/MiddlewareTests.swift [hunk 7]
```diff
@@ -85,46 +82,37 @@ class MiddlewareTests: XCTestCase {
...
-
-        // test to make sure basic server saw the
-        // header the middleware added
-        XCTAssert(try res.bodyString().contains("Foo") == true)
-        XCTAssert(try res.bodyString().contains("bar") == true)
-
-        // test to make sure the middleware
-        // added headers to the response
-        XCTAssertEqual(res.headers["bar"], "baz")
+//        var config = Config([:])
```
Comment (LoganWright) [inline Tests/VaporTests/MiddlewareTests.swift]: rm or uncomment 👍

File: Sources/Vapor/Engine/ClientProtocol.swift (hunk 4 above)
Comment (tanner0101) [inline Sources/Vapor/Engine/ClientProtocol.swift]: This allows us to do `drop.client.defaultTLSContext = ...` which has better discoverability than `Global.defaultTLSClientContext`

File: Tests/VaporTests/DropletTests.swift (hunk 6 above)
Comment (tanner0101) [inline Tests/VaporTests/DropletTests.swift]: ah yeah, i meant to add that back in. I was trying to think of a way to automate it (maybe in the type safe routing package).

I found an error where one of them was missing `= []`. maintaining tons of copies is hard.

File: Sources/Vapor/Droplet/Droplet+TLS.swift (hunk 1 above)
Comment (tanner0101) [inline Sources/Vapor/Droplet/Droplet+TLS.swift]: `.port` converts an `Int` to a `UInt16`. Maybe I should just make the call `.uint16`

File: Tests/VaporTests/DropletTests.swift (hunk 6 above)
Comment (tanner0101) [inline Tests/VaporTests/DropletTests.swift]: I added back all of these overloads.

**Pull request merged**